framework/include/constraints/ADMortarScalarBase.h

```diff
@@ -0,0 +1,88 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "ADMortarConstraint.h"
+
+/**
+ * This Constraint adds standardized methods for assembling to a primary
+ * scalar variable associated with the major variables of the Mortar
+ * Constraint object. Essentially, the entire row of the residual and Jacobian
+ * associated with this scalar variable will also be assembled here
+ * using the loops over mortar segments.
+ * This variable is "scalar_variable" in the input file and "kappa"
+ * within the source code.
+ */
+
+class ADMortarScalarBase : public ADMortarConstraint
+{
+public:
+  static InputParameters validParams();
+
+  ADMortarScalarBase(const InputParameters & parameters);
+
+  // Using declarations necessary to pull in computeResidual with different parameter list and avoid
+  // hidden method warning
+  using ADMortarConstraint::computeResidual;
+
+  // Using declarations necessary to pull in computeJacobian with different parameter list and avoid
+  // hidden method warning
+  using ADMortarConstraint::computeJacobian;
+
+  /**
+   * The scalar variable that this kernel operates on.
+   */
+  const MooseVariableScalar & scalarVariable() const { return *_kappa_var_ptr; }
+
+  /**
+   * Computes _var-residuals as well as _kappa-residual
+   */
+  virtual void computeResidual() override;
+  /**
+   * Computes d-_var-residual / d-_var and d-_var-residual / d-jvar,
+   * as well as d-_kappa-residual / d-_var and d-_kappa-residual / d-jvar
+   */
+  virtual void computeJacobian() override;
+
+protected:
+  /**
+   * Method for computing the scalar part of residual at quadrature points
+   */
+  virtual ADReal computeScalarQpResidual() { return 0; }
+
+  /**
+   * Put necessary evaluations depending on qp but independent of test functions here
+   */
+  virtual void initScalarQpResidual() {}
+
+  /// Whether a scalar variable is declared for this constraint
+  const bool _use_scalar;
+
+  /// Whether to compute scalar contributions for this instance
+  const bool _compute_scalar_residuals;
+
+  /// A dummy object useful for constructing _kappa when not using scalars
+  const ADVariableValue _kappa_dummy;
```

Comment:
I don't know whether it might make sense to use `_ad_zero` from `Coupleable` instead?

Reply:
I was using this logic of dummy variables because I saw it in the constructor of `MortarConstraint` and `ADMortarConstraint` for the Lagrange multiplier `_lambda`. Similar to the conversation above, I view kappa like lambda in the mortar constraint file, namely that if the user doesn't set the "variable" parameter then they don't intend to deal with lower variables but rather secondary/primary variables. Is the default constructor of a variable equal to 0? I'm looking at `Coupleable.h` and seeing both `_zero` and `_ad_zero`, which I could use for the non-AD and AD files if that is desirable.
Maybe should there be other assert conditions elsewhere to catch unintended usage? Like the comment after the constructor that says "add some error checks here"?

Reply:
It compiled with `_zero` and `_ad_zero` and it compiled and the tests that use the constructor from zero still ran. So these will make the code simpler, and have been incorporated.